### PR TITLE
infidel Sensor with Table in EEPROM and Programming over I2C

### PR DESCRIPTION
**Pull request by midopple**
Hi,

I have make an Firmware for the Sensor with
- Table in EEPROM
- Programm the Table over I2C (Host)
- Get Diameter
- Get RAW Value

so the Sensor has Programmed one time an the Calibartion kan make over I2C with Uart Console on the Host.

Michael

### CODE REVIEW

**Comment by midopple** [comment]
Analog out is also available 
Voltage = Diamater

**Comment by drspangle** [comment]
This is very nice. I would like to accept this PR after there is a bit more documentation about how to use it, and once it is more complete. An installation guide and test procedure should be written so people will know it is working correctly (I assume this is the same as the existing firmware?). It will become confusing for people if they do not know which firmware files to use at least.

**Comment by midopple** [comment]
HI,
I try to make an documentation  about the Firmware with installing wireing and testing to make ist easier

**Comment by midopple** [comment]
Hi,

I added the Wire Diagramm and the first documentation, iam not the best English writer so i hobe it is useable

**Comment by drspangle** [comment]
@midopple Wenn Sie eine Dokumentation in Deutsch schreiben möchten... Das ist ebenfalls akzeptabel ! Ich kann gleich übersetzen.

Ein mögliches Problem: Dies kann zu Funktionsabweichungen von der Hauptfirmware führen. Wir müssen dies für einige Zeit in einer speziellen Branche aufbewahren. Vielleicht nur vorübergehend.  Ist das OK für Ihnen?

**Comment by midopple** [comment]
@drspangle 

I have completed the documentation in English, explaining the firmware and also the process of calibrating via the I2C.
See the readme.me

In my opinion it is not necessary to make a separate branch for the change because my new version is compatible to the first version, the sensor can be addressed with the original software where only 2 bytes are queried via the I2C. 
The functions such as calibrating via I2C or analogue output are additional and are addressed via their own commands, after switching on the sensor behaves as in the first program version.

One consideration is whether to put the firmware in its own folder.

If the sensor is already connected to a printer, the new firmware works without problems.

Greetings
Michael

**Comment by drspangle** [comment]
This is good work Michael, thank you for your valuable contribution. Your effort is appreciated very much.

I will take some time to review everything so that we can merge this. We will both need to think carefully about the best way to approach merging. I can see why there would be a benefit to including both versions (the original and your alternative) firmwares. However we should ensure that there is one source of documentation only. I can potentially work on merging between the README.md you created and the original file. I will also do some light editing to correct any small errors that may be found since I am a native English speaker.

**Comment by midopple** [comment]
No problem, they were right that I speak German. Thank you for checking the document and correcting the mistakes.

File: firmware/Host_ee_prog.ino [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+
+  Wire.begin();        // join i2c bus (address optional for host)
+  Serial.begin(19200);  // start serial for output
+
+  Serial.println(F("Infidel Sensor Programmer"));
+
+  //Check if the Device at Adress is Online
+  check_I2C_adress();
+   
+  Serial.println(F("Command Input (0 - val / 1 - RAW val / 2 - Version / 3 - Table / 4 - Set Tabel Val / 5 - Ongoing raw read):"));
```
Comment (coelholm) [inline firmware/Host_ee_prog.ino:43]: Serial.println(F("Command Input (0 - val / 1 - RAW val / 2 - Version / 3 - Table / 4 - Set Table Val / 5 - Ongoing raw read):"));

File: firmware/Host_ee_prog.ino [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+      case '5':
+        if(ongoing_read == 0)
+          ongoing_read = 1;
+        else
+          ongoing_read = 0;
+      break;
+      case 'h':
+      case 'H':
+        Serial.println(F("Commands:"));
+        Serial.println(F("Command Input (0 - val / 1 - RAW val / 2 - Version / 3 - Table / 4 - Set Tabel Val / 5 - Ongoing raw read)"));
```
Comment (coelholm) [inline firmware/Host_ee_prog.ino:87]: Serial.println(F("Command Input (0 - val / 1 - RAW val / 2 - Version / 3 - Table / 4 - Set Table Val / 5 - Ongoing raw read)"));

File: firmware/README.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+```sh
+Version: 1.11
+Table [ADC] [DIA in um]:
+00: 0001 / 2999
+01: 0617 / 2092
+02: 0722 / 1711
+03: 0816 / 1401
+04: 0999 / 1001
+05: 1022 / 0001
+Command Input (0 - val / 1 - RAW val / 2 - Version / 3 - Table / 4 - Set Tabel Val):
```
Comment (coelholm) [inline firmware/README.md:62]: Command Input (0 - val / 1 - RAW val / 2 - Version / 3 - Table / 4 - Set Table Val):

File: firmware/README.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+note the ADC Value and use the Command "4" 
+The Console show 
+
+```sh
+Input values for Table [IDX],[ADC],[DIA um] like (1,619,2090)
+Input: 
+```
+
+Input this String --> 1,503,2000
+Means, Tabel Index 1 (Comnand "3"), ADC Val 503, Diameter 3
```
Comment (coelholm) [inline firmware/README.md:90]: Means, Table Index 1 (Comnand "3"), ADC Val 503, Diameter 3

**Comment by midopple** [comment]
Thanks for the Info, i have changed the text in readme.md and host_ee_prog

**Comment by coelholm** [comment]
I have used version #8 and it is great work, IMHO.  I have calibrated diameter following the procedure and I also had to update fromLow and fromHigh values at map command to reflect real voltage output when analogWrite is set to 0 and 255 respectively. I forced duty cycle manually and update firmware in order to see voltage output and pin OUT, but I believe it worth create a new command to have it forced 0 and 255 for 5 or 10 seconds each, so we can get both output. This way I can implement another command via i2c at device and add it to prog host and make calibration process even easier, a command to save values to EEPROM would help as well.

**Review by drspangle — APPROVED** [review]

**Comment by drspangle** [comment]
> 
> 
> I have used version #8 and it is great work, IMHO. I followed calibration procedure and it was pretty simple, but I also had to update fromLow and fromHigh values at map command to reflect real voltage output when analogWrite is set to 0 and 255 respectively. In my sensor i2c diameter in mm was different from the voltage output. In order to calibrate output I forced duty cycle manually (0 and 255) to check voltage at pin OUT, but I believe it worth create a new command to force it 0 and 255 for 5 or 10 seconds each, so we can get both outputs. This way I can implement another command via i2c at device and access it from prog host and make calibration process even easier, a command to save/read those values to/from EEPROM would help as well.

Please create a new with a feature request for this.

**Pull request merged**